mono/metadata/mono-perfcounters.c

```diff
@@ -1483,276 +1567,438 @@ typedef struct {
+	/* emit the SharedCounter structures */
+	*p++ = perfctr_type_compress (MONO_HANDLE_GETVAL (data, type));
+	*p++ = i;
+	// FIXME This should be UTF8 conversion into supplied buffer, previously
+	// sized correctly.
+	p = append (p, counter_info [i * 2]);
+	p = append (p, counter_info [i * 2 + 1]);
+
+	HANDLE_FUNCTION_RETURN_VAL (p);
+}
+
 /*
  * Since we'll keep a copy of the category per-process, we should also make sure
  * categories with the same name are compatible.
  */
 MonoBoolean
-mono_perfcounter_create (MonoString *category, MonoString *help, int type, MonoArray *items)
+ves_icall_System_Diagnostics_PerformanceCounterCategory_Create (
+	MonoStringHandle category, MonoStringHandle help, int type, MonoArrayHandle items, MonoError *error)
 {
-	MonoError error;
+// FIXME integer overflow throughout this function.
+// FIXME This function unnecessarily builds up all the data
+// into a temporary, and then copies it to shared memory.
+// There is very little value in the temporary it is just extra code and cost.
+// All the temporary does is avoid sizing utf8 data separate from conversion to utf8.
+// Sizing and copying are roughly the same cost (strlen vs. strcpy), and this
+// "optimization" costs heap.
+
+	error_init (error);
+
 	int result = FALSE;
-	int i, size;
-	int num_counters = mono_array_length (items);
-	int counters_data_size;
+	int const num_counters = mono_array_handle_length (items);
 	char *name = NULL;
 	char *chelp = NULL;
 	char **counter_info = NULL;
-	char *p;
-	SharedCategory *cat;
+	gboolean unlock = FALSE;
 
 	/* FIXME: ensure there isn't a category created already */
-	name = mono_string_to_utf8_checked (category, &error);
-	if (!mono_error_ok (&error))
-		goto failure;
-	chelp = mono_string_to_utf8_checked (help, &error);
-	if (!mono_error_ok (&error))
-		goto failure;
+	name = mono_string_handle_to_utf8 (category, error);
+	goto_if_nok (error, failure);
+
+	chelp = mono_string_handle_to_utf8 (help, error);
+	goto_if_nok (error, failure);
+	
+	// Build up all the data into a temporary and then copy
+	// into shared memory. FIXME avoid the extra copying.
+
 	counter_info = g_new0 (char*, num_counters * 2);
-	/* calculate the size we need structure size + name/help + 2 0 string terminators */
-	size = G_STRUCT_OFFSET (SharedCategory, name) + strlen (name) + strlen (chelp) + 2;
+	if (!counter_info) // FIXME out of memory handling
+		goto failure;
+
+	int i;
 	for (i = 0; i < num_counters; ++i) {
-		CounterCreationData *data = mono_array_get (items, CounterCreationData*, i);
-		counter_info [i * 2] = mono_string_to_utf8_checked (data->name, &error);
-		if (!mono_error_ok (&error))
-			goto failure;
-		counter_info [i * 2 + 1] = mono_string_to_utf8_checked (data->help, &error);
-		if (!mono_error_ok (&error))
-			goto failure;
-		size += sizeof (SharedCounter) + 1; /* 1 is for the help 0 terminator */
+		ves_icall_System_Diagnostics_PerformanceCounterCategory_Create_create_temp (
```

Comment:
One thing that we do sometimes for loop body functions like this is to change their return type to `gboolean` so that they return with `HANDLE_FUNCTION_RETURN_VAL (is_ok (error))` and then you can write

```c
for (...) {
  if (!helper (..., error))
    goto failure;
}
```

Reply:
Ah, I did think briefly here before settling on void, if there'd be something useful to return, but didn't think of that. But I like the one line instead of two error handling at the caller..though what you say is more efficient.. Really anything other than exceptions is too verbose.